querier/src/cache/object_store.rs

```diff
@@ -0,0 +1,465 @@
+    }
+
+    async fn list_with_delimiter(
+        &self,
+        prefix: Option<&Path>,
+    ) -> Result<ListResult, ObjectStoreError> {
+        self.inner.list_with_delimiter(prefix).await
+    }
+
+    async fn copy(&self, _from: &Path, _to: &Path) -> Result<(), ObjectStoreError> {
+        Err(ObjectStoreError::NotImplemented)
+    }
+
+    async fn copy_if_not_exists(&self, _from: &Path, _to: &Path) -> Result<(), ObjectStoreError> {
+        Err(ObjectStoreError::NotImplemented)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use assert_matches::assert_matches;
+    use futures::TryStreamExt;
+    use iox_time::SystemProvider;
+    use metric::{Attributes, DurationHistogram, Metric};
+    use object_store::memory::InMemory;
+    use object_store_metrics::ObjectStoreMetrics;
+
+    use crate::cache::ram::test_util::test_ram_pool;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test() {
+        // set up inner store with content
+        let inner = Arc::new(InMemory::new());
+
+        let path_1 = Path::from("foo");
+        let bytes_1 = Bytes::from(b"data_foo" as &'static [u8]);
+        inner.put(&path_1, bytes_1.clone()).await.unwrap();
+
+        let path_2 = Path::from("bar/1");
+        let bytes_2 = Bytes::from(b"data_bar/1" as &'static [u8]);
+        inner.put(&path_2, bytes_2.clone()).await.unwrap();
+
+        let path_3 = Path::from("bar/2");
+        let bytes_3 = Bytes::from(b"data_bar/2" as &'static [u8]);
+        inner.put(&path_3, bytes_3.clone()).await.unwrap();
+
+        let path_4 = Path::from("baz");
+        let bytes_4 = Bytes::from(b"data_baz" as &'static [u8]);
+
+        // set up cache
+        let metric_registry = metric::Registry::new();
+        let time_provider = Arc::new(SystemProvider::new());
+        let instrumented_store = ObjectStoreMetrics::new(
+            Arc::clone(&inner) as _,
+            Arc::clone(&time_provider) as _,
+            &metric_registry,
+        );
+        let cache = ObjectStoreCache::new(
+            BackoffConfig::default(),
+            Arc::new(instrumented_store),
+            time_provider,
+            &metric_registry,
+            test_ram_pool(),
+            true,
+        );
+        let cached_store = cache.object_store();
+
+        // ensure "hits" are cached
+        assert_eq!(get_count_hit(&metric_registry), 0);
+        assert_eq!(
+            cached_store
+                .get(&path_1)
+                .await
+                .unwrap()
+                .bytes()
+                .await
+                .unwrap(),
+            bytes_1,
+        );
+        assert_eq!(get_count_hit(&metric_registry), 1);
+        assert_eq!(
+            cached_store
+                .get(&path_1)
+                .await
+                .unwrap()
+                .bytes()
+                .await
+                .unwrap(),
+            bytes_1,
+        );
+        assert_eq!(get_count_hit(&metric_registry), 1);
+
+        // ensure "misses" are cached
```

Comment:
```suggestion
        // ensure "not found" are cached
```
Too many overloads :laughing: